fish-rust/src/env/environment.rs

```diff
@@ -371,8 +379,315 @@
+                            EnvMode::GLOBAL | EnvMode::EXPORT,
+                            str2wcstring(s.to_bytes()),
+                        );
+                    } else {
+                        vars.set_empty(L!("HOME"), EnvMode::GLOBAL | EnvMode::EXPORT);
+                    }
+                }
+                return;
+            }
+        }
+    }
+
+    // Either we didn't have a $USER or it had a different uid.
+    // We need to get the data *again* via the uid.
+    let retval = unsafe {
+        libc::getpwuid_r(
+            uid,
+            userinfo.as_mut_ptr(),
+            buf.as_mut_ptr(),
+            buf.len(),
+            &mut result,
+        )
+    };
+    if retval == 0 && !result.is_null() {
+        let userinfo = unsafe { userinfo.assume_init() };
+        let s = unsafe { CStr::from_ptr(userinfo.pw_name) };
+        let uname = str2wcstring(s.to_bytes());
+        vars.set_one(L!("USER"), EnvMode::GLOBAL | EnvMode::EXPORT, uname);
+        // Only change $HOME if it's empty, so we allow e.g. `HOME=(mktemp -d)`.
+        // This is okay with common `su` and `sudo` because they set $HOME.
+        if vars.get_unless_empty(L!("HOME")).is_none() {
+            if !userinfo.pw_dir.is_null() {
+                let s = unsafe { CStr::from_ptr(userinfo.pw_dir) };
+                vars.set_one(
+                    L!("HOME"),
+                    EnvMode::GLOBAL | EnvMode::EXPORT,
+                    str2wcstring(s.to_bytes()),
+                );
+            } else {
+                // We cannot get $HOME. This triggers warnings for history and config.fish already,
+                // so it isn't necessary to warn here as well.
+                vars.set_empty(L!("HOME"), EnvMode::GLOBAL | EnvMode::EXPORT);
+            }
+        }
+    } else if vars.get_unless_empty(L!("HOME")).is_none() {
+        // If $USER is empty as well (which we tried to set above), we can't get $HOME.
+        vars.set_empty(L!("HOME"), EnvMode::GLOBAL | EnvMode::EXPORT);
+    }
+}
+
+/*
+fn setup_path() {
+    let vars = &GLOBALS;
+    let path = vars.get_unless_empty(L!("PATH"));
+    if path.is_none() {
+        todo!("confstr and _CS_PATH are not available in Rust libc");
+        vars.set_one(
+            L!("PATH"),
+            EnvMode::GLOBAL | EnvMode::EXPORT,
+            "/usr/bin:/bin".into(),
+        );
+    }
+}
+*/
+
+pub fn env_init(paths: Option<&ConfigPaths>, do_uvars: bool, default_paths: bool) {
+    let vars = &PRINCIPAL_STACK;
+
+    let env_iter: Vec<_> = std::env::vars_os()
+        .map(|(k, v)| (str2wcstring(k.as_bytes()), str2wcstring(v.as_bytes())))
+        .collect();
+    // Import environment variables. Walk backwards so that the first one out of any duplicates wins
+    // (See issue #2784).
+    // PORTING: this should behave like in C++ in regards to the above comment, but that might be
+    // with assumptions on the standard library's behavior.
+    for (key, val) in env_iter.into_iter().rev() {
```

Comment:
right. Maybe we should call `environ()` ourselves

Reply:
probably we should keep walking environ backwards, unless we know that there is no security issue here

Reply:
This is walking backwards, it calls `.rev()`, the std-implementation walks forwards, as in https://github.com/rust-lang/rust/blob/2ceed0b6cb9e9866225d7cfcfcbb4a62db047163/library/std/src/sys/unix/os.rs#L560, and AFAIK this did not fix any known vulnerabilities when it was added, ref. https://github.com/fish-shell/fish-shell/issues/2784